src/java.base/share/classes/jdk/internal/classfile/impl/Util.java

```diff
@@ -321,4 +322,73 @@ interface Writable {
     interface WritableLocalVariable {
         boolean writeLocalTo(BufWriterImpl buf);
     }
+
+    /**
+     * Returns the hash code of an internal name given the class or interface L descriptor.
+     */
+    public static int internalNameHash(String desc) {
+        if (desc.length() > 0xffff)
+            throw new IllegalArgumentException("String too long (" + desc.length() + "): " + desc);
+        return (desc.hashCode() - pow31(desc.length() - 1) * 'L' - ';') * INVERSE_31;
+    }
+
+    // k is at most 65536, length of Utf8 entry + 1
+    public static int pow31(int k) {
+        int r = 1;
+        // calculate the power contribution from index-th octal digit
+        // from least to most significant (right to left)
+        // e.g. decimal 26=octal 32, power(26)=powerOctal(2,0)*powerOctal(3,1)
+        for (int i = 0; i < SIGNIFICANT_OCTAL_DIGITS; i++) {
+            r *= powerOctal(k & 7, i);
+            k >>= 3;
+        }
+        return r;
+    }
+
+    // The inverse of 31 in Z/2^32Z* modulo group, a * INVERSE_31 * 31 = a
+    static final int INVERSE_31 = 0xbdef7bdf;
+
+    // k is at most 65536 = octal 200000, only consider 6 octal digits
+    // Note: 31 powers repeat beyond 1 << 27, only 9 octal digits matter
+    static final int SIGNIFICANT_OCTAL_DIGITS = 6;
+
+    // for base k, storage is k * log_k(N)=k/ln(k) * ln(N)
+    // k = 2 or 4 is better for space at the cost of more multiplications
+    /**
+     * The code below is as if:
+     * {@snippet lang=java :
+     * int[] powers = new int[7 * SIGNIFICANT_OCTAL_DIGITS];
+     *
+     * for (int i = 1, k = 31; i <= 7; i++, k *= 31) {
+     *    int t = powers[powersIndex(i, 0)] = k;
+     *    for (int j = 1; j < SIGNIFICANT_OCTAL_DIGITS; j++) {
+     *        t *= t;
+     *        t *= t;
+     *        t *= t;
+     *        powers[powersIndex(i, j)] = t;
+     *    }
+     * }
+     * }
+     * This is converted to explicit initialization to avoid bootstrap overhead.
```

Comment:
Have you measured this to be true and useful? Static array initializers are bulky at a bytecode level so sometimes just generating on the fly carries insignicant overhead. Then the `UtilTest` assertion test wouldn't really be needed.

Reply:
I measured in bytestacks and the clinit instructions reduced by about 2/3

Reply:
Yes, you have to pick and evaluate between two trade-offs here: a compact routine to compute the array, or a larger initializer which does less compute. Number of instructions executed in the interpreter is a diagnostic tool since if often hides costs that come from having to store, read, parse and manage bytecode - sometimes twice if this hits the default CDS archive. There's no definite answer, but generally computers have trended towards IO = expensive, compute = cheap, so picking a larger representation to avoid some one-off compute might be the wrong trade-off.

Reply:
I will leave this hash computation mechanism whole as is; table initialization and small table optimizations can come in other patches later.